src/components/Tokens/TokenDetails/index.tsx

```diff
@@ -64,6 +64,28 @@ function useOnChainToken(address: string | undefined, skip: boolean) {
   }
 }
 
+// Selects most relevant token based on data available, preferring native > query > on-chain
+// Token will be null if still loading from on-chain, and undefined if unavailable
+function useRelevantToken(
+  address: string | undefined,
+  pageChainId: number,
+  tokenQueryData: TokenQueryData | undefined
+) {
+  const { chainId: activeChainId } = useWeb3React()
+  const queryToken = useMemo(() => {
+    if (!address) return undefined
+    if (address === NATIVE_CHAIN_ID) return nativeOnChain(pageChainId)
+    if (tokenQueryData) return new QueryToken(tokenQueryData)
+    return undefined
+  }, [pageChainId, address, tokenQueryData])
+  // fetches on-chain token if query data is missing and page chain matches global chain (else fetch won't work)
+  const skipOnChainFetch = Boolean(queryToken) || pageChainId !== activeChainId
+  const onChainToken = useOnChainToken(address, skipOnChainFetch)
+  const token = useMemo(() => queryToken ?? onChainToken, [queryToken, onChainToken])
```

Comment:
This should not be memoized - it's only a boolean check, and it does not change the reference. It could also be included inline in the returned object.

    const token = queryToken ?? onChainToken